test/ResolutionManager.ts

```diff
@@ -132,61 +99,54 @@ describe("Resolution", async () => {
       .whenCalledWith(RESOLUTION_ROLE, deployer.address)
       .returns(true);
 
-    await voting.mock_getDelegateAt(user1.address, user1.address);
-
-    // Set contributors' status to:
-    // - contributor
-    // - shareholder
-    // - investor
-    // The mock is dumb so we need to set everything manually
-    await Promise.all(
-      [managingBoard, user1, user2, delegate1].map(
-        async (user) => await setContributor(user, true)
-      )
-    );
-    await shareholderRegistry.mock_isAtLeast(
-      managingBoardStatus,
-      managingBoard.address,
-      true
-    );
-
-    await Promise.all(
-      [managingBoard, user1, user2, delegate1].map(async (voter) => {
-        await voting.mock_canVoteAt(voter.address, true);
-      })
-    );
+    shareholderRegistry.snapshot.returns(resolutionSnapshotId);
+    voting.snapshot.returns(resolutionSnapshotId);
+    token.snapshot.returns(resolutionSnapshotId);
   });
 
   beforeEach(async () => {
     snapshotId = await network.provider.send("evm_snapshot");
-    resolutionId = 1; // first resolution ID is always 1
+    shareholderRegistry.isAtLeast
+      .whenCalledWith(managingBoardStatus, managingBoard.address)
+      .returns(true);
   });
 
   afterEach(async () => {
     await network.provider.send("evm_revert", [snapshotId]);
+    shareholderRegistry.isAtLeast.reset();
+    shareholderRegistry.isAtLeastAt.reset();
+    token.balanceOf.reset();
+    token.balanceOfAt.reset();
+    voting.canVoteAt.reset();
+    voting.getDelegate.reset();
+    voting.getDelegateAt.reset();
+    voting.getVotingPower.reset();
+    voting.getVotingPowerAt.reset();
+    voting.getTotalVotingPower.reset();
+    voting.getTotalVotingPowerAt.reset();
   });
 
-  async function setContributor(user: SignerWithAddress, flag: boolean) {
-    await shareholderRegistry.mock_isAtLeast(
-      contributorStatus,
-      user.address,
-      flag
-    );
-    await shareholderRegistry.mock_isAtLeast(
-      shareholderStatus,
-      user.address,
-      flag
-    );
-    await shareholderRegistry.mock_isAtLeast(
-      investorStatus,
-      user.address,
-      flag
-    );
-  }
-
-  async function setupUser(user: SignerWithAddress, votingPower: number) {
-    await voting.mock_getDelegateAt(user.address, user.address);
-    await voting.mock_getVotingPowerAt(user.address, votingPower);
+  function setupUser(
+    user: SignerWithAddress,
+    votingPower: number,
+    balance = 0,
```

Comment:
Not sure I understand why we need both `votingPower` and `balance`. In which case they differ?

Reply:
internally they have their specific uses: when someone delegates and then overrides their power of attorney, we use the balance instead of the voting power to count the total votes for that contributor.